### ISSUE TITLE
Using CONTAINS via Gremlin script

### ISSUE THREAD

**Issue by zachkinstner** [thinkaurelius/titan#272]
I'm having issues using `CONTAINS` in Gremlin queries. I'm using RexConnect, so these queries are coming in via RexPro -- I'm not building these queries in a Java program.

``` c
g.v(123).has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x")
```

```
com.tinkerpop.rexster.client.RexProException> 
An error occurred while processing the script for language [groovy]. 
All transactions across all graphs in the session have been concluded with failure: 
javax.script.ScriptException: 
groovy.lang.MissingMethodException: 
No signature of method: 
com.tinkerpop.gremlin.groovy.GremlinGroovyPipeline.has() is applicable for argument types: (java.lang.String, com.thinkaurelius.titan.core.attribute.Text$1, java.lang.Integer) 
values: [MyValue, CONTAINS, 1]
Possible solutions: has(java.lang.String, java.lang.Object), has(java.lang.String, com.tinkerpop.gremlin.Tokens$T, java.lang.Object), cap(), last(), max(), as(java.lang.String)
```

After looking at more examples, I was able to do this:

``` c
g.query().has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x").vertices()
```

Will it be possible to use `CONTAINS` beyond the initial index? For example:

``` c
g.v(123).outE.inV.has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x")
```

Also -- I didn't see the full `com.thinkaurelius.titan.core.attribute.Text.CONTAINS` path noted anywhere. Is this the expected way to do it via Gremlin script, or am I missing some kind of import/include?

**Comment by zachkinstner** [comment]
I did find a note from Marko [here](https://groups.google.com/forum/?fromgroups=#!topic/gremlin-users/XJX2hOAvyOY):

> We need to start putting such concepts into the Blueprints layer so graph engines can take advantage of them.

**Comment by mbroecheler** [comment]
Hey Zach,

text and geo search capabilities are currently limited to Titan. As you
noted, we hope to bring them into Blueprints into the near future to enable
the query you were attempting to run.

As such, this is not really a Titan issue but rather a Blueprints/Gremlin
issue.

On Tue, May 14, 2013 at 6:46 PM, Zach Kinstner notifications@github.comwrote:

> I'm having issues using CONTAINS in Gremlin queries. I'm using
> RexConnect, so these queries are coming in via RexPro -- I'm not building
> these queries in a Java program.
> 
> g.v(123).has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x")
> 
> com.tinkerpop.rexster.client.RexProException>
> An error occurred while processing the script for language [groovy].
> All transactions across all graphs in the session have been concluded with failure:
> javax.script.ScriptException:
> groovy.lang.MissingMethodException:
> No signature of method:
> com.tinkerpop.gremlin.groovy.GremlinGroovyPipeline.has() is applicable for argument types: (java.lang.String, com.thinkaurelius.titan.core.attribute.Text$1, java.lang.Integer)
> values: [MyValue, CONTAINS, 1]
> Possible solutions: has(java.lang.String, java.lang.Object), has(java.lang.String, com.tinkerpop.gremlin.Tokens$T, java.lang.Object), cap(), last(), max(), as(java.lang.String)
> 
> After looking at more examples, I was able to do this:
> 
> g.query().has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x").vertices()
> 
> Will it be possible to use CONTAINS beyond the initial index? For example:
> 
> g.v(123).outE.inV.has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x")
> 
> Also -- I didn't see the full
> com.thinkaurelius.titan.core.attribute.Text.CONTAINS path noted anywhere.
> Is this the expected way to do it via Gremlin script, or am I missing some
> kind of import/include?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/272
> .

## 

Matthias Broecheler
http://www.matthiasb.com

**Comment by spmallette** [comment]
zach, is it possible to get around the problem by casting `g` to `TitanGraph` before doing `query()`?

**Comment by mbroecheler** [comment]
No, the has() after g.V. is an optimization in Gremlin. Without CONTAINS
support in gremlin that optimization will not be possible.

On Wed, May 15, 2013 at 4:07 AM, stephen mallette
notifications@github.comwrote:

> zach, is it possible to get around the problem by casting g to TitanGraphbefore doing
> query()?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/272#issuecomment-17932336
> .

## 

Matthias Broecheler
http://www.matthiasb.com

**Comment by spmallette** [comment]
oh...right...got confused.

**Comment by zachkinstner** [comment]
I did try to cast `g`, with no luck:

``` c
# RexConnect> query
#   ...script (0; string): ((com.thinkaurelius.titan.core.TitanGraph)g).v(175920).has('x', com.thinkaurelius.titan.core.attribute.Text.CONTAINS, 'y')
#   ...params (1; string; opt): 

{"timer":644,"err":"com.tinkerpop.rexster.client.RexProException> An error occurred while processing the script for language [groovy]. All transactions across all graphs in the session have been concluded with failure: javax.script.ScriptException: groovy.lang.MissingMethodException: No signature of method: groovy.lang.MissingMethodException.has() is applicable for argument types: (java.lang.String, com.thinkaurelius.titan.core.attribute.Text$1, java.lang.String) values: [x, CONTAINS, y]\nPossible solutions: wait(), any(), find(), dump(), grep(), is(java.lang.Object)"}
```

So, the conclusion seems to be: updated Blueprints/Gremlin support is coming in the near future.

In the meantime, two questions... First, will this example query (noted above, executes correctly) actually make use of ElasticSearch when called via RexConnect/RexPro?

``` c
g.query().has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x").vertices()
```

Second, am I doing something incorrectly to require the full path: `com.thinkaurelius.titan.core.attribute.Text.CONTAINS`?  I've found I also have to use full paths for other statics/constants, like [TypeGroup.of](https://github.com/inthefabric/Fabric/blob/master/Solution/Fabric.Db.Data/Setups/SetupIndexesFuncs.cs#L15) and [NO_LOCK](https://github.com/inthefabric/Fabric/blob/master/Solution/Fabric.Db.Data/Setups/SetupIndexesFuncs.cs#L26).

**Comment by mbroecheler** [comment]
On the second issue: I know that the gremlin shell that ships with Titan
includes those enums in the imports, but I believe that might not be the
case for Rexster by default, right @spmallette?

On Wed, May 15, 2013 at 1:01 PM, Zach Kinstner notifications@github.comwrote:

> I did try to cast g, with no luck:
> 
> # RexConnect> query#   ...script (0; string): ((com.thinkaurelius.titan.core.TitanGraph)g).v(175920).has('x', com.thinkaurelius.titan.core.attribute.Text.CONTAINS, 'y')#   ...params (1; string; opt):
> 
> {"timer":644,"err":"com.tinkerpop.rexster.client.RexProException> An error occurred while processing the script for language [groovy]. All transactions across all graphs in the session have been concluded with failure: javax.script.ScriptException: groovy.lang.MissingMethodException: No signature of method: groovy.lang.MissingMethodException.has() is applicable for argument types: (java.lang.String, com.thinkaurelius.titan.core.attribute.Text$1, java.lang.String) values: [x, CONTAINS, y]\nPossible solutions: wait(), any(), find(), dump(), grep(), is(java.lang.Object)"}
> 
> So, the conclusion seems to be: updated Blueprints/Gremlin support is
> coming in the near future.
> 
> In the meantime, two questions... First, will this example query (noted
> above, executes correctly) actually make use of ElasticSearch when called
> via RexConnect/RexPro?
> 
> g.query().has("MyValue", com.thinkaurelius.titan.core.attribute.Text.CONTAINS, "x").vertices()
> 
> Second, am I doing something incorrectly to require the full path:
> com.thinkaurelius.titan.core.attribute.Text.CONTAINS? I've found I also
> have to use full paths for other statics/constants, like TypeGroup.ofhttps://github.com/inthefabric/Fabric/blob/master/Solution/Fabric.Db.Data/Setups/SetupIndexesFuncs.cs#L15and
> NO_LOCKhttps://github.com/inthefabric/Fabric/blob/master/Solution/Fabric.Db.Data/Setups/SetupIndexesFuncs.cs#L26
> .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/thinkaurelius/titan/issues/272#issuecomment-17962457
> .

## 

Matthias Broecheler
http://www.matthiasb.com

**Comment by spmallette** [comment]
Yes...that is not the case with Rexster.  I had a discussion with marko on this the other day and it needs more though.  Rexster uses the GremlinGroovyScriptEngine to execute gremlin.  Since it does not depend on Titan it does not do titan imports.  Follow this issue over in Rexster:

https://github.com/tinkerpop/rexster/issues/304

**Comment by mbroecheler** [comment]
Issue has been closed in Rexster which closes it in Titan for the next release (0.4.0)
